### PR TITLE
Item-Change: Validate NBT & that stack size is greater than rollback. Handle partial stacks 

### PR DESCRIPTION
**Pull request by Altirix**
Changes:

ItemChangeActionType.kt 

- Check Item NBT matches the item being removed (and when added to partial stacks of same type)
- can now iteratively remove itemStacks that are less than the rollback/restore itemStack size
- Add to existing itemStacks that are of the same item & nbt

- IDEA detects a `SENSELESS_COMPARISON` within this logic, however, removing this check changes the logic

SlotMixin.java 

- logChange is no longer a recursive function & certain checks have been reordered & done as else to prior checks
- add or remove to stack of same type will retain their NBT/ other custom properties
- removal of redundant casting between PlayerEntity and ServerPlayerEntity

- needs validation that is functions exactly the same as before as its a fairly major change

ScreenHandlerMixin.java

- check if the index of the slot clicked is EMPTY_SPACE_SLOT_INDEX (-999) which is used for non-slot operations adding this stopped me from being able to desync player sources on transactions however why this exactly works is unknown to me. it might not even fully fix the issue but im yet to find a test case.

Unresolved issues:

- there's no way to check before a rollback/restore if any actions will fail (this will only occur if there's no item to remove or no space to add items, so this should be quite rare), however, this can result in item loss or excess items being added to containers when chained with other actions. there is no fix once this has happened. its mostly an issue when certain transactions are not logged (hoppers)

- SlotMixin.java oldStack is not updated when another player has changed the contents when 2+ players are in the same container. unsure how to proceed with this just yet  https://streamable.com/v2zi19

### CODE REVIEW

File: src/main/kotlin/com/github/quiltservertools/ledger/actions/ItemChangeActionType.kt [hunk 1]
```diff
@@ -71,37 +71,63 @@ abstract class ItemChangeActionType : AbstractActionType() {
...
-        if (world != null && inventory != null) {
-            val rollbackStack = ItemStack.fromNbt(StringNbtReader.parse(extraData))
-
-            for (i in 0 until inventory.size()) {
-                val stack = inventory.getStack(i)
-                if (stack.isItemEqual(rollbackStack)) {
-                    inventory.setStack(i, ItemStack.EMPTY)
-                    return true
-                }
+        if (world == null || inventory == null) { return false }
```
Comment (yitzy299) [inline src/main/kotlin/com/github/quiltservertools/ledger/actions/ItemChangeActionType.kt:74]: I'd prefer a check for if not null

**Review by yitzy299 — COMMENTED** [review]
As with #127, the whole PR could do with a run through the formatter

**Comment by Altirix** [comment]
The changes to ScreenHandler have been removed, when shift-clicking items into a container the index is of the player inventory

**Comment by DrexHD** [comment]
Superseded by #278

**Pull request closed without merging**